astropy/nddata/utils.py

```diff
@@ -600,23 +609,48 @@ def __init__(self, data, position, shape=None, side_length=None, wcs=None,
                                  'SkyCoord')
             position = skycoord_to_pixel(position, wcs, mode='all')  # (x, y)
 
-        if side_length is None and shape is None:
-            raise ValueError("Either side_length or shape must be specified")
-
-        if side_length is not None and shape is not None:
-            raise ValueError("Cannot specify both side_length and shape")
-
-        if side_length is not None:
-            shape = (side_length, side_length)
-
-        shape = [x.value if u.pixel.is_equivalent(x) else x for x in shape]
-
-        # extract_array and overlap_slices use (y, x) positions
-        pos = position[::-1]
+        if np.isscalar(size):
+            size = np.repeat(size, 2)
+
+        # special handling for a scalar Quantity
+        if isinstance(size, u.Quantity):
+            size = np.atleast_1d(size)
+            if len(size) == 1:
+                size = np.repeat(size, 2)
+
+        if len(size) > 2:
+            raise ValueError('size must have at most two elements')
+
+        shape = np.zeros(2).astype(int)
+        pixel_scales = None
+        # ``size`` can have a mixture of int and Quantity (and even units),
+        # so evaluate each axis separately
+        for axis, side in enumerate(size):
+            if not isinstance(side, u.Quantity):
+                shape[axis] = size[axis]     # pixels
+            else:
+                if side.unit is u.pixel:
+                    shape[axis] = side.value
+                elif side.unit.physical_type == 'angle':
+                    if wcs is None:
+                        raise ValueError('wcs must be input if any element '
+                                         'of size has angular units')
+                    if pixel_scales is None:
+                        pixel_scales = u.Quantity(
+                            proj_plane_pixel_scales(wcs), wcs.wcs.cunit[axis])
```

Comment:
This may very well not be a concern in any cases, but how much can pixel scale vary across an image? From the `proj_plane_pixel_scales` docs, that function calculates pixel scale only at the image center...